### PR TITLE
[move][move-ide] Revise how IDE information is recorded in the compiler

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

This PR moves how we record IDE information to communicate it to the Move Analyzer. It is now accrued on the `CompilationEnv` (modulo some care in typing to handle type elaboration) and can be added through methods there. The goal is to set up an extensible way to track IDE information in the future, tying it to location information.

This change has also caused us to change how we compute spans for dotted expression terms: previously, we would make the expression with the call's span in order to report failures to autoborrow to point at the call site (instead of just the subject term). This caused recurring through location information in the analyzer to enter an infinite loop in the macro case, as the subject term shared the location of the overall macro call and recursion into the subject term on the macro information looped forever. This diff introduces a solution to this issue by more-carefully computing spans for subject terms so that they remain pointing at the actual subject term, and plumbs the call location through dotted expression handling to retain good error reporting for the method case. It also introduces a very small change to handle location reporting for `*`/`&`/`&mut ` usage for type errors.

## Test plan 

All tests still pass.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/shared/ide.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+#[derive(Debug, Clone)]
+pub struct IDEInfo {
+    pub exp_info: BTreeMap<Loc, ExpEntry>
+}
+
+#[derive(Debug, Clone)]
+pub struct ExpEntry {
+    /// Location of the recorded info
+    pub loc: Loc,
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/shared/ide.rs:18]: Is this a different location than the one used as key of the `exp_info` map (e.g., indicates a different, more precise location inside of the expression)?

File: external-crates/move/crates/move-compiler/src/typing/translate.rs [hunk 2]
```diff
@@ -3769,12 +3765,12 @@ fn method_call(
                 edotted.for_autocomplete = true;
                 let err_ty = context.error_type(loc);
                 let dot_output =
-                    resolve_exp_dotted(context, DottedUsage::Borrow(false), loc, edotted);
+                    resolve_exp_dotted(context, DottedUsage::Borrow(false), edotted.loc, edotted);
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/typing/translate.rs:3768]: Is this related or unrelated fix?

File: external-crates/move/crates/move-compiler/src/shared/ide.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /// Indicates this location was a macro call site
+    pub macro_call_info: Option<Box<MacroCallInfo>>,
+    /// Indicates this location is an expansion of a lambda
+    pub expanded_lambda: bool,
+}
+
+/// For use in the compiler, to ease recording IDE information.
+pub enum ExpInfo {
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/shared/ide.rs:26]: It looks like `ExpEntry` and `ExpInfo` encode the same type of information. Can we just use one rather than using `ExpInfo` in the compiler and (at least seemingly) translating it almost immediately into `ExpEntry`?

**Review by awelc — COMMENTED** [review]
Left some comments and some cleanup would be necessary but overall looks good

File: external-crates/move/crates/move-compiler/src/shared/ide.rs (hunk 1 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/shared/ide.rs:21]: No, it's just duplicated here for sanity.

File: external-crates/move/crates/move-compiler/src/typing/translate.rs (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/typing/translate.rs]: All of the span stuff is quite related. I'll provide a write-up in the mian PR when I publish it for review.

File: external-crates/move/crates/move-compiler/src/shared/ide.rs (hunk 3 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/shared/ide.rs:29]: Unfortunately, this is likely the cleanest thing we can do. Since the actual information live on the `CompilationEnv` take a `Diagnostics` to report ICEs, the only alternative would be to re-expose the _entire_ interface for every information form on that environment, e.g., 

```
pub fun set_ide_exp_macro_call_info(&mut self, call_info: ide::MacroCallInfo) {
  ...
}

pub fun set_ide_exp_expanded_lambda(&mut self) {
  ...
}

pub fun set_ide_exp_autocomplete_dot_information(&mut self) {
  ...
}
```

The current approach avoids this by essentially baking this dispatch into the `ExpInfo`. I'm happy to change this if you want, but I think the overall usage and maintenance will become clumsier.

File: external-crates/move/crates/move-compiler/src/shared/ide.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    shared::Name, typing::ast as T,
+};
+
+use move_ir_types::location::Loc;
+
+use std::collections::BTreeMap;
+
+#[derive(Debug, Clone)]
+pub struct IDEInfo {
+    pub exp_info: BTreeMap<Loc, ExpEntry>,
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/shared/ide.rs:15]: I'm a bit worried about the longevity of the design here, as it forces a good handful of assumptions onto the structure of `ExpEntry`.

I would probably prefer something like `info: Vec<(Loc, ExpInfo)>`, and let the IDE sort it out. 
The mental model I have here is that the compiler can blindly tag locations with information, then the IDE can come along and sort them all out. 
This isn't too different to what we do with Diags, we have a big vector we push to, that we then muck around with when it comes time to report

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 5]
```diff
@@ -55,6 +55,7 @@
 #![allow(clippy::non_canonical_partial_ord_impl)]
 
 use crate::{
+    compiler_info::CompilerInfo,
```
Comment (awelc) [inline external-crates/move/crates/move-analyzer/src/symbols.rs:58]: Is this module missing?

**Review by tnowacki — APPROVED** [review]
